design/parallel-commit/globally-non-unique-timestamps.md

```diff
@@ -1,3 +1,27 @@
+# Allow commit_ts to be non-globally unique.
+
+This document is about the problem and the solution of the non-globally-unique `commit_ts` problem.
+
+## The problem
+
+in TiKV, rolling back a transaction need to leave a rollback record in order to guarantee consistency. But rollback records and commit records are both saved in Write CF, and the commit records are saved with `{user_key}{commit_ts}` as the internal key, while that of rollback records are `{user_key}{start_ts}`. Previously the `commit_ts`es are TSOs allocated from PD, which is guaranteed globally unique. However when we tries to use a calculated timestamp as the `commit_ts` to avoid the latency of PD's RPC, it's possible that a rollback record gets the same internal key as a commit record, but we need to keep them both.
+
+## The solution
+
+The solution is to keep the commit record, but with a `has_overlapped_rollback` flag in this case to indicate that there's a rollback happened here whose `start_ts` equals to the current record's `commit_ts`. But only "protected" rollbacks need to set the flag. Non-protected rollback records can be dropped without introducing potential inconsistency.
+
+To do this, when performing a protected rollback operation, check the records in write CF to see if there's already a commit record of another transaction. If so, instead of writing the rollback record, add the `has_overlapped_rollback` flag to that commit record. For example, when rolling a transaction whose `start_ts = 10` on a key, but there's another commit record whose `start_ts = 5` and `commit_ts = 10` that updates the key, if we write the rollback record directly, then the commit record will be overwritten. So instead, keep the commit record, but set the `has_overlapped_rollback` flag.
+
+It's also possible that when committing a transaction, there's already another transaction's rollback record that might be overwritten, which is not expected. An easy approach to solve this is to check whether there's a overlapping rollback record already here. But previously commit operations don't need to read Write CF, so introducing this check may significantly harm the performance. Our final solution to this case is that:
```

Comment:
```suggestion
It's also possible that when committing a transaction, there's already another transaction's rollback record that might be overwritten, which is not expected. An easy approach to solve this is to check whether there's a overlapping rollback record already here. But before async commit is introduced, commit operations didn't need to read Write CF, so introducing this check may significantly harm the performance. Our final solution to this case is that:
```

Reply:
It might be better to explain it a bit more: commit is in the hot path while rollbacks happen far less often, so the following approach is better